src/pages/[platform]/build-a-backend/data/connect-to-existing-data-sources/single-table-design/index.mdx

```diff
@@ -0,0 +1,490 @@
+
+To interact with the external DynamoDB table, you can add custom types, mutations, and queries to the `amplify/data/resource.ts` file.
+
+Here's an example schema for the racing application:
+
+```ts
+import { type ClientSchema, a, defineData } from "@aws-amplify/backend";
+
+const schema = a.schema({
+  Race: a.customType({
+    raceId: a.string().required(),
+    classId: a.string(),
+    results: a.json(),
+  }),
+  RacingTable: a.customType({
+    pk: a.string().required(),
+    sk: a.string().required(),
+    numeric: a.float(),
+    results: a.json(),
+  }),
+  createRace: a
+    .mutation()
+    .arguments({
+      raceId: a.string().required(),
+      classId: a.string(),
+      results: a.json(),
+    })
+    .returns(a.ref("RacingTable"))
+    .handler(
+      a.handler.custom({
+        dataSource: "RacingTableDataSource",
+        entry: "./createRace.js",
+      })
+    )
+    .authorization((allow) => [allow.publicApiKey()]),
+  getRaceResultsByRacerId: a
+    .query()
+    .arguments({
+      racerId: a.string().required(),
+    })
+    .returns(a.ref("RacingTable").array())
+    .handler(
+      a.handler.custom({
+        dataSource: "RacingTableDataSource",
+        entry: "./getRaceResultsByRacerId.js",
+      })
+    )
+    .authorization((allow) => [allow.publicApiKey()]),
+  getRacesByClassId: a
+    .query()
+    .arguments({
+      classId: a.string().required(),
+    })
+    .returns(a.ref("RacingTable").array())
+    .handler(
+      a.handler.custom({
+        dataSource: "RacingTableDataSource",
+        entry: "./getRacesByClassId.js",
+      })
+    )
+    .authorization((allow) => [allow.publicApiKey()]),
+  getBestPerformanceByClassId: a
+    .query()
+    .arguments({
+      classId: a.string().required(),
+    })
+    .returns(a.ref("RacingTable").array())
+    .handler(
+      a.handler.custom({
+        dataSource: "RacingTableDataSource",
+        entry: "./getBestPerformanceByClassId.js",
+      })
+    )
+    .authorization((allow) => [allow.publicApiKey()]),
+});
+
+export type Schema = ClientSchema<typeof schema>;
+
+export const data = defineData({
+  schema,
+  name: "MyLibrary",
+  authorizationModes: {
+    defaultAuthorizationMode: "apiKey",
+    apiKeyAuthorizationMode: {
+      expiresInDays: 365,
+    },
+  },
+});
+```
+
+## Reviewing Data Access Patterns
+
+It's crucial to test the proposed schema against the list of data access patterns. Let's revisit the racing application's queries to ensure the table schema supports them:
+
+1. **Get results for each race by racer ID**: Use the table's partition key, searching for PK = racer ID. This returns a list of all races (PK) for a specific racer.
+
+<BlockSwitcher>
+<Block name="Resolver">
+```ts title="amplify/data/getRaceResultsByRacerId.js"
+import { util } from "@aws-appsync/utils";
+import * as ddb from "@aws-appsync/utils/dynamodb";
+
+export function request(ctx) {
+  const { racerId } = ctx.args;
+
+  return ddb.query({
```

Comment:
I couldn't find any information about a query command. I tried it and it doesn't seem to work.

> "message": "code.js:7:10: Property 'query' is expected to be a valid condition object",
> I/flutter ( 8117):     "locations": [
> I/flutter ( 8117):       {
> I/flutter ( 8117):         "line": 2,
> I/flutter ( 8117):         "column": 7
> I/flutter ( 8117):       }
> I/flutter ( 8117):     ],
> I/flutter ( 8117):     "path": [
> I/flutter ( 8117):       "querySuE"
> I/flutter ( 8117):     ],
> I/flutter ( 8117):     "errorType": "Code"

Reply:
Strange. .query should be a valid method based on the AppSync docs:

https://docs.aws.amazon.com/appsync/latest/devguide/tutorial-dynamodb-resolvers-js.html#configure-query

I didn't run into issues deploying or invoking it in a request from the client side

Reply:
My mistake... Sry
It works!